CLI: using pnpm causes multiple `npm list` errors written to CLI output for `provider`, `synth` and `debug` commands

### Expected Behavior

When the `cdktf-cli` package is installed as a dev dependency via `pnpm`, no `npm` errors are written to the output.

### Actual Behavior

When running some commands (`provider get`, `debug`, `synth`, multiple errors are written of the following form:

```log
[2023-05-31T23:22:01.376] [ERROR] default - npm
npm[2023-05-31T23:22:01.377] [ERROR] default -  ERR! code ELSPROBLEMS
npm ERR! invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs

 ERR! code ELSPROBLEMS
npm ERR! invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs
[2023-05-31T23:22:01.377] [ERROR] default -
npm ERR! A complete log of this run can be found in:
npm

npm ERR! A complete log of this run can be found in:
npm[2023-05-31T23:22:01.378] [ERROR] default -  ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_21_59_841Z-debug-0.log

 ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_21_59_841Z-debug-0.log
```

I'm uncertain if this is causing an issue or not: the command ends up succeeding.

### Steps to Reproduce

1. Create a typescript CDKTF project
2. Convert to the pnpm package manager (`pnpm import yarn.lock` or `pnpm import package-lock.json`)
3. `pnpm add cdktf-cli@latest`
4. Run `pnpm install`
5. Run `pnpm exec cdktf provider get` or `pnpm exec cdktf debug`

### Versions

```log
> [project]-cdk@0.1.0 cdktf /Users/[user]/src/[project]/ops
> cdktf "debug"

[2023-05-31T23:18:21.003] [ERROR] default - npm
npm[2023-05-31T23:18:21.004] [ERROR] default -  ERR! code ELSPROBLEMS
npm ERR! invalid: jsii@5.0.10 /Users/[user]/src/[project]/ops/node_modules/.pnpm/node_modules/jsii

 ERR! code ELSPROBLEMS
npm ERR! invalid: jsii@5.0.10 /Users/[user]/src/[project]/ops/node_modules/.pnpm/node_modules/jsii
[2023-05-31T23:18:21.005] [ERROR] default -


[2023-05-31T23:18:21.005] [ERROR] default - npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log
[2023-05-31T23:18:21.006] [ERROR] default - npm
npm[2023-05-31T23:18:21.006] [ERROR] default -  ERR! code ELSPROBLEMS

 ERR! code ELSPROBLEMS
[2023-05-31T23:18:21.007] [ERROR] default - npm ERR!
npm ERR![2023-05-31T23:18:21.007] [ERROR] default -  invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs

 invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs
[2023-05-31T23:18:21.008] [ERROR] default -


[2023-05-31T23:18:21.008] [ERROR] default - npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log
cdktf debug
language: typescript
cdktf-cli: 0.16.1
node: v18.16.0
cdktf: 0.16.1
constructs: null
jsii: null
terraform: 1.4.6
arch: arm64
os: darwin 22.4.0
```

### Providers

```text
┌───────────────┬──────────────────┬─────────┬────────────┬─────────────────────┬─────────────────┐
│ Provider Name │ Provider Version │ CDKTF   │ Constraint │ Package Name        │ Package Version │
├───────────────┼──────────────────┼─────────┼────────────┼─────────────────────┼─────────────────┤
│ aws           │ 4.66.1           │ ^0.16.0 │            │ @cdktf/provider-aws │ 14.0.4          │
└───────────────┴──────────────────┴─────────┴────────────┴─────────────────────┴─────────────────┘
```

### Gist

https://gist.github.com/jamestelfer/03f844d1e1dc83df2c7d79730545a7a3

### Workarounds

I haven't seen this with a globally installed CLI, but I'm not 100% certain why.

We don't install the CLI globally as we want the local development environment consistent with CI, and we have multiple projects that don't necessarily change CLI versions at the same time.

### Anything Else?

The command being run (`npm list jsii --json` or similar) works with pnpm: `pnpm list jsii --json`. I haven't checked the JSON output.

If `pnpm` was run if a `pnpm-lock.yaml` is detected, this would avoid the error.

I don't know what the effect of this error is.

I'm willing to have a look at a fix, but I'm unsure of the timeline that I'd have to contribute it -- I might be able to get someone on my team to have a look.

### References

_No response_

### Help Wanted

- [X] I'm interested in contributing a fix myself

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

--- a/packages/@cdktf/commons/src/debug.ts
+++ b/packages/@cdktf/commons/src/debug.ts
@@ -6,6 +6,7 @@ import { logger } from "./logging";
 import { exec } from "./util";
 import { terraformVersion } from "./terraform";
 import { DISPLAY_VERSION } from "./version";
+import { pathExists } from "fs-extra";
 
 export function getLanguage(projectPath = process.cwd()): string | undefined {
   try {
@@ -73,7 +74,59 @@ export function getNodeVersion() {
 async function getNodeModuleVersion(
   packageName: string
 ): Promise<string | undefined> {
+  // Use the presence of the pnpm lock file as a signal that
+  // we should interrogate packages via pnpm instead of npm.
+  const usingPnpm = await pathExists("pnpm-lock.yaml");
+
+  return usingPnpm
+    ? getPnpmNodeModuleVersion(packageName)
+    : getNpmNodeModuleVersion(packageName);
+}
+
+async function getPnpmNodeModuleVersion(
+  packageName: string
+): Promise<string | undefined> {
   let output;
+
+  try {
+    output = await exec("pnpm", ["list", packageName, "--json"], {
+      env: { ...process.env },
+    });
+  } catch (e) {
+    logger.debug(`Unable to run 'pnpm list ${packageName} --json': ${e}`);
+    return undefined;
+  }
+
+  let json;
+  try {
+    json = JSON.parse(output);
+  } catch (e) {
+    logger.debug(
+      `Unable to parse output of 'pnpm list ${packageName} --json': ${e}`
+    );
+    return undefined;
+  }
+
+  if (
+    !json ||
+    !Array.isArray(json) ||
+    json.length === 0 ||
+    !json[0]?.dependencies?.[packageName]?.version
+  ) {
+    logger.debug(
+      `Unable to find '${packageName}' in 'npm list ${packageName} --json': ${output}`
+    );
+    return undefined;
+  }
+
+  return json[0].dependencies[packageName].version;
+}
+
+async function getNpmNodeModuleVersion(
+  packageName: string
+): Promise<string | undefined> {
+  let output;
+
   try {
     output = await exec("npm", ["list", packageName, "--json"], {
       env: { ...process.env },

--- a/packages/@cdktf/commons/src/debug.ts
+++ b/packages/@cdktf/commons/src/debug.ts
@@ -114,7 +114,7 @@ async function getPnpmNodeModuleVersion(
     !json[0]?.dependencies?.[packageName]?.version
   ) {
     logger.debug(
-      `Unable to find '${packageName}' in 'npm list ${packageName} --json': ${output}`
+      `Unable to find '${packageName}' in 'pnpm list ${packageName} --json': ${output}`
     );
     return undefined;
   }

